### ISSUE TITLE
Support table and column comments modeled in Prisma

### ISSUE THREAD

**Issue by bill-coulam** [prisma/prisma#8703]
## Problem

The relational database engines I am familiar with (Oracle, postgres, SQL Server, etc.) all have an internal "data dictionary" that contains metadata about all the schema and data objects, like tables, views, materialized views, indexes, columns, etc. They all allow comments on these objects to be declared using DDL statements such as:

`COMMENT ON COLUMN mytable.mycolumn IS 'This is what this column holds. Here is the history. Here is a caveat and a weird case. etc.'`

Prisma does not support documenting the schema in the database's internal data dictionary.

Data definitions are critical. I can't begin to express how important this is for future engineers on a given project, for data engineers, report writers, data scientists, InfoSec, legal, data catalog/glossary tools and efforts, governance, etc.

## Suggested solution

Add an attribute to the Prisma modeling spec, something like @comment("Here is my awesome comment about this data item.") that can be placed to the right of each column, or parameterized version like @comment("my_table_name", "Here is the full business definition of this table and how it is used.") for a table or enum. When Prisma finally supports views and materialized views, those should allow comments as well.

Take each @comment string and generate and run a comment DDL statement to insert it into the underlying DB data dictionary, right after creating the table or data object. If the underlying DB engine doesn't support comments, then Prisma would forego generating and running the comment statements. In this case, the definitions in the Prisma model could still be useful for the team and data catalog tools.

## Alternatives

As of now, my teams have to write manual DDL scripts, and run them after Prisma has done its model migration into the DB schema. It is not ideal and clunky.

## Additional context

If I weren't in a hurry, I'd investigate all the databases for which you have built native connectors, to see if which ones support data dictionary comments.

**Comment by yi-ge** [comment]
good idea!

**Comment by janpio** [comment]
Potentially related issues: https://github.com/prisma/prisma/issues/4818 + https://github.com/prisma/prisma/issues/8896

**Comment by StringKe** [comment]
Does anyone know if the implementation of `@comment` needs to change the `datamodel` => `transform` related code in the `prisma-engines` warehouse?

ref: https://github.com/prisma/prisma-engines/tree/master/libs/datamodel/core/src/transform/ast_to_dml/db/attributes

**Comment by yi-ge** [comment]
@StringKe I am also trying to do this, +weixin?

**Comment by StringKe** [comment]
> @StringKe I am also trying to do this, +weixin?

https://github.com/prisma/prisma-engines/pull/2737 

The more basic functionality has been implemented within this PR.
已经在这个 PR 内实现了比较基础的功能。

**Comment by janpio** [comment]
Prisma Schema Language already supports comment that are added to the internal data structures with triple slash comments: https://www.prisma.io/docs/concepts/components/prisma-schema#comments What does not happen yet, is that we introspect comments from databases and render them into triple slash comment, or that we persist triple slash comments to tables or columns when migrating the Prisma schema to a database schema.

The suggested `@comment` attribute per model or field would clash with that, so an alternative solution might be to just reuse these existing comments to store the same information. Or is there a reason why this might be a bad idea? 

(On the other hand, we might also replace triple slash comments with `@comment` instead as it is a potentially better way to connect comments to their "objects")

**Comment by StringKe** [comment]
I recommend using the `@comment` method.

`@comment` is a comment in the database
`///` is to generate typescript type comments

**Comment by janpio** [comment]
That would be very explicit 👍 What if you want these to be the same though? Shouldn't database descriptions for example make their way into the Prisma Client after an initial `prisma db pull`?

**Comment by xiangnanscu** [comment]
@janpio strongly recommend using `@comment` method! The `///` is implicit.

**Comment by StringKe** [comment]
> That would be very explicit 👍 What if you want these to be the same though? Shouldn't database descriptions for example make their way into the Prisma Client after an initial `prisma db pull`?

Yes, the in-database `comment` should be added to the `prisma schema` on initialization

**Comment by raarts** [comment]
I'm very afraid I'll end up writing two comment lines for each field, one for TypeScript, and one for the database documentation. Experience tells me, that in practice these two are quite often the same. And in exceptional cases can be worded such, that they're useful in both contexts.

So how about a global directive: generate_db_comments_from_slash_comments (please come up with a better name). And the occasional @comment directive to override this on a per field/table basis.

**Comment by bill-coulam** [comment]
I agree with @raarts
Best if a data object or field is described only once, and a global directive tells Prisma whether to also generate an in-database comment, with the option to override if for some odd reason the description in the database needed to read differently than the one in typescript.

**Comment by boblauer** [comment]
> I'm very afraid I'll end up writing two comment lines for each field, one for TypeScript, and one for the database documentation. Experience tells me, that in practice these two are quite often the same. And in exceptional cases can be worded such, that they're useful in both contexts.

If you have the comment inline in a `@comment` method, why would you also need the TS comment?

**Comment by raarts** [comment]
@boblauer I was responding to @StringKe , who said:

> I recommend using the @comment method.
> 
> @comment is a comment in the database
> /// is to generate typescript type comments

The TS comments are presumably used for generating comments in IDE's TS help.
To which I said: they serve the same purpose and I expect them generally to be the same.

**Comment by raarts** [comment]
I checked and indeed the triple slash comments end up in the index.d.ts file of the generated code:

prisma file:
```
model Event {
  id            Int       @id @default(autoincrement()) /// unique id for this event
```

generated typescript:
```
export type Event = {
  /**
   * unique id for this event 
   */
  id: number
```

**Comment by boblauer** [comment]
Interesting, maybe it wasn't working for me because i had the comment on it's own line above the column.

If the @comment function also added a TS comment that would get rid of the need for 2 separate comments.

**Comment by raarts** [comment]
True, but /// comments are already accepted, have an infrastructure behind them, and are used for other purposes in various plugins. Moving to @comments would not be backwards compatible.

And to me, using triple slashes already are intuitive and organic.

**Comment by janpio** [comment]
(No need to continue this discussion, we are aware of the different potential or real approaches to define comments in Prisma Schema and will consider both when we look into resolving this issue. Right now we have no further opinion than that triple slash comments are already implement to solve another use case.)

**Comment by MaximusMcCann** [comment]
PLEASE 🙏

**Comment by markthepixel** [comment]
Really need this, it's such a pain without it. Save us time please!!!!!

**Comment by macoto1995** [comment]
Hi. @janpio 
We really need this as an application development team to collaborate with our data management team.
DB comment inside DB itself is the best way to communicate about what is the column.
Please accept this issue:pray:

**Comment by fruwe** [comment]
This would be really helpful to have 🥺 🙏

**Comment by tunguyennnnn** [comment]
this would be super super helpful!!!

**Comment by harshitshah436** [comment]
I would love to see this feature.

**Comment by unstoppable-allensun** [comment]
Can this please be revisted? Our team could really use this feature.

**Comment by sudoryan** [comment]
I think this is a great idea and would really save me a lot of time

**Comment by unstoppable-allensun** [comment]
Can this please be revisted? Our team could really use this feature.

**Comment by helfi92** [comment]
+1

**Comment by masatada-kurihara** [comment]
pls

**Comment by Mr-Geneo** [comment]
+1

**Comment by ggebi** [comment]
+1

**Comment by dalkommatt** [comment]
+1

**Comment by fabioselau077** [comment]
+1

**Comment by Oxicode** [comment]
anything news?

**Comment by oishiimendesu** [comment]
👀

**Comment by gristow** [comment]
This issue is also a challenge for us. Because Prisma is not the only way we interact with our database, we would prefer to rely on comments within the database as the single source of truth for schema documentation.

I am wondering whether writing a prisma generator might be a possibility for getting this functionality?  I have not taken a close look at the generator API, so, my real question is: in theory, could a generator introspect a database for comments, and add those to the jsdoc style comments that prisma outputs in `prisma generate`?

(Alternately, I suppose a standalone program could be written to parse the prisma schema file, query the database for comments, and then add them to the schema file. This feels much less clean to me, however.)

**Comment by StringKe** [comment]
> This issue is also a challenge for us. Because Prisma is not the only way we interact with our database, we would prefer to rely on comments within the database as the single source of truth for schema documentation.
> 
> I am wondering whether writing a prisma generator might be a possibility for getting this functionality? I have not taken a close look at the generator API, so, my real question is: in theory, could a generator introspect a database for comments, and add those to the jsdoc style comments that prisma outputs in `prisma generate`?
> 
> (Alternately, I suppose a standalone program could be written to parse the prisma schema file, query the database for comments, and then add them to the schema file. This feels much less clean to me, however.)

I have implemented a similar feature if you need it, you can check this pr to maintain a version yourself.

https://github.com/prisma/prisma-engines/pull/2737

Also I have uploaded a parsing support `@stringke/mmg` which you can use to get the parsed results and generate your own sql

**Comment by gristow** [comment]
Thanks, @StringKe. We wouldn't want to rely on a non-supported version of the core prisma engine, or a non-supported change to the prisma schema language. But if I have some time I may look at whether it would be possible to use some of your work to instead create a [prisma generator](https://prismaio.notion.site/Prisma-Generators-a2cdf262207a4e9dbcd0e362dfac8dc0) that would do this instead.

**Comment by Jyrno42** [comment]
A very rudimentary generator that I added into my project that fits our use-case. May be useful for others too.

```
#! /usr/bin/env ts-node

/**
 * This is a custom generator for Prisma that generates comments for all models fields and
 *  handles creating a migration file for them when comments change.
 * 
 * The comments are generated from the documentation field in the Prisma schema (e.g. the /// comments
 *  in the schema file).
 * 
 * It works based on a lock file of all comment statements. When it detects that the comments have 
 * changed (by comparing the sha256 hash of them), the commited lock file will be updated. In addition,
 * a new migration file will be created with all comments.
 * 
 * For our purposes its not a big issue since running the sql statements that add comments
 *  should be cheap anyway.
 * 
 * This is a workaround to have https://github.com/prisma/prisma/issues/8703 before it is implemented
 * in Prisma itself.
 */

import { promises as fs } from 'fs';
import { createHash } from 'crypto';
import debug from 'debug';

import { parseEnvValue, getDMMF } from '@prisma/internals';
import { EnvValue, GeneratorOptions, generatorHandler } from '@prisma/generator-helper';

const debugLog = debug('prisma:generate-comments');

async function generateModelComment(model: any): Promise<string[]> {

    const modelName = model.dbName ?? model.name;

    const commentStatements: string[] = [];
    
    model.fields.forEach((field: any) => {
        if (!field.documentation) {
            return;
        }

        debugLog(`Generating comment for ${modelName}.${field.name}...`);

        const escapedComment = field.documentation?.replace(/'/g, "''") ?? '';

        const commentTemplate = `COMMENT ON COLUMN "${modelName}.${field.name}" IS '${escapedComment}';`;
        commentStatements.push(commentTemplate);
    });

    return [
        `-- Model ${modelName} comments`,
        '',
        ...commentStatements,
        ''
    ];
};

async function fileHash(file: string, allowEmpty = false): Promise<string> {
    try {
        const fileContent = await fs.readFile(file, 'utf-8');

        // now use sha256 to hash the content and return it
        return createHash('sha256').update(fileContent).digest('hex');
    } catch (e: any) {
        if (e.code === 'ENOENT' && allowEmpty) {
            return '';
        }

        throw e;
    }
}

async function lockChanged(lockFile: string, tmpLockFile: string): Promise<boolean> {
    return await fileHash(lockFile, true) !== await fileHash(tmpLockFile);
} 

export async function generate(options: GeneratorOptions) {
  const outputDir = parseEnvValue(options.generator.output as EnvValue);
  await fs.mkdir(outputDir, { recursive: true });

  const prismaClientProvider = options.otherGenerators.find(
    (it) => parseEnvValue(it.provider) === 'prisma-client-js',
  );

  const prismaClientDmmf = await getDMMF({
    datamodel: options.datamodel,
    previewFeatures: prismaClientProvider?.previewFeatures,
  });

  const promises: Promise<string[]>[] = [];

  prismaClientDmmf.datamodel.models.forEach((model: any) => {
    debugLog(`Generating comment for ${model.name}...`);
    promises.push(generateModelComment(model));
  });

  const allStatements = await Promise.all(promises);

  const tmpLock = await fs.open(`${outputDir}/.comments-lock.tmp`, 'w+');

  await tmpLock.write('-- generator-version: 1.0.0\n\n');

  // concat all promises and separate with new line and two newlines between each model
  const allStatementsString = allStatements.map((statements) => statements.join('\n')).join('\n\n');

  await tmpLock.write(allStatementsString);
  await tmpLock.close();

  // compare hashes of tmp lock file and existing lock file
  // if they are the same, do nothing
  // if they are different, write tmp lock file to lock file
  // if lock file does not exist, also write tmp lock file to lock file
  const isChanged = await lockChanged(`${outputDir}/.comments-lock`, `${outputDir}/.comments-lock.tmp`);

  if (isChanged) {
    await fs.copyFile(`${outputDir}/.comments-lock.tmp`, `${outputDir}/.comments-lock`);

    // when lockfile changed we generate a new migration file too
    const date = new Date();
    date.setMilliseconds(0);

    const dateStr = date.toISOString().replace(/[:\-TZ]/g, '').replace('.000', '');
    const migrationDir = `prisma/migrations/${dateStr}_update_comments`;

    console.log(`Lock file changed, creating a new migration at ${migrationDir}...`);

    await fs.mkdir(migrationDir, { recursive: true });

    await fs.copyFile(`${outputDir}/.comments-lock`, `${migrationDir}/migration.sql`);
  } else {
    console.log('No changes detected, skipping creating a fresh comment migration...');
  }

  // always delete tmp lock file
  await fs.unlink(`${outputDir}/.comments-lock.tmp`);

  console.log('Comment generation completed');
}


generatorHandler({
  onManifest() {
    return {
      defaultOutput: 'comments',
      prettyName: 'Prisma Database comments Generator',
    };
  },
  onGenerate: generate,
});
```

**Comment by janpio** [comment]
@gristow Generally, a generator only gets some information from the schema, and then can run any code it wants and use that information there. I don't know if that is what you are looking for. Introspection (that reads the database to get information) is a different system all together for example.

**Comment by woss** [comment]
@Jyrno42 thanks so much for this solution, this was the critical part to enable almost-full functionality of postgraphile.

For others who might want to know how to use this here is a quick overview of how I use it:
1. save the snippet in the `src/commentsGenerator.ts` 
2. remove shebang (`#! /usr/bin/env ts-node`)  if you have local ts-node and not global
3. add this to your `schema.prisma`. this assumes you have `ts-node` in package.json
```gql
generator comments {
  provider = "ts-node src/commentsGenerator.ts"
}
```
4. add the scripts to the package.json
```jsonc
{
  // ...
  "scripts": {
    // ...
    "dev:deploy": "prisma migrate deploy",
    "dev:migrate": "prisma migrate dev && pnpm dev:deploy"
    // ...
  }
  // ...
}
```
5.  now you can on development just run it like `pnpm dev:migrate` and it will add the comments as well

**Comment by Jyrno42** [comment]
You're welcome @woss. If you (or anyone else) wants to turn this into a Gist or an actual package then feel free. All I ask is that maybe I can get an honorary mention somewhere (and you send a link to the thing here so I can change my project to use it too down the line).

**Comment by woss** [comment]
@Jyrno42 i already have it as a package in my project. if there is an interest I will spend a bit more time and publish it as cjs and esm flavors. 

OFC you get the mention and link to the source code as I already did in the file I have :) 
<img width="705" alt="Screenshot 2023-07-14 at 16 39 05" src="https://github.com/prisma/prisma/assets/80440/99f7eaba-3691-45f8-80b6-a29f9369346b">

**Comment by EloB** [comment]
What is the chances of this feature reaching core? I model my database with Prisma but I consume it using [postgraphile](https://postgraphile.org/) and they use "smart tags". This feature would really fit that :)

**Comment by janpio** [comment]
Very high, but there is no timeline for when that happens.

**Comment by EloB** [comment]
@wl1092212424 Did you mange to get `@@schema` to work with this script? I can't find `schema` in the script.

**Comment by samislam** [comment]
Any updates on this?

**Comment by zC4sTr0** [comment]
Any updates on this issue? Seems like there's a PR https://github.com/prisma/prisma-engines/pull/2737 on this already
This feature would be much appreciated

**Comment by divmgl** [comment]
Yep, in need of PostgreSQL schema comments to clearly delineate when a column should be used over another (they're both JSON columns).

**Comment by Mateusnasciment** [comment]
alguma atualização ?

**Comment by bakhatyar** [comment]
Hey there, Prisma team! 🎉

Wow, your article on GitHub issue #8703 shed some light on the importance of maintaining clarity and consistency within the Prisma documentation. As a developer who frequently relies on Prisma for database management in my projects, I can't stress enough how crucial it is to have accurate and well-organized documentation. It's like having a reliable map to navigate through complex territories, and your efforts in addressing this issue directly contribute to smoother sailing for developers like myself.

I particularly resonated with your emphasis on providing comprehensive examples and ensuring that the documentation caters to developers across various skill levels. It's reassuring to see that you're actively listening to the community's feedback and striving to make the documentation not just informative but also accessible and user-friendly. Kudos to your team for being proactive in addressing concerns and continuously enhancing the developer experience. Looking forward to seeing the positive impact these improvements will have on the Prisma ecosystem. 
Keep up the fantastic work! 🚀
<a href="https://truthtablegenerator.net/">truth table generator</a>

**Comment by johannbuscail** [comment]
Any update @janpio ?

**Comment by janpio** [comment]
No, otherwise we would have posted an update here @johannbuscail.

**Comment by bakhatyar** [comment]
Hey there, ! Just finished reading your article, and I gotta say, it hit the nail right on the head. Your insights into the challenges developers face with schema migrations are spot on. I've been in the trenches myself, wrestling with database changes and trying to keep everything in sync. It's a headache, to say the least.

But your suggestion about leveraging Prisma's --create-only flag? Genius. Seriously, that little nugget of wisdom is a game-changer. I can already picture myself using it to streamline my workflow and avoid those pesky data loss scenarios. And your breakdown of the issue on GitHub? Clear as day. It's reassuring to know that others out there are tackling the same hurdles and finding innovative solutions like this. Thanks for sharing your expertise with the community!
<a href="https://pictionarywordgenerator.com/">Pictionary Word Generator</a>

**Comment by normancarcamo** [comment]
Thank you authors for such a great job, I'm in love with Prisma!

But... comments are necessary for DBs though, I hope this can be taken into consideration in the future because they are needed.

**Comment by Lililili-li** [comment]
Please improve this feature

**Comment by onozaty** [comment]
I created it as an npm package based on https://github.com/prisma/prisma/issues/8703#issuecomment-1614360386 idea.
Please try it if you like.

- https://github.com/onozaty/prisma-db-comments-generator

A big thank you to @Jyrno42

**Comment by kelvinCJJ** [comment]
Cheers mate, please take a look

> I created it as an npm package based on [#8703 (comment)](https://github.com/prisma/prisma/issues/8703#issuecomment-1614360386) idea. Please try it if you like.
> 
> * https://github.com/onozaty/prisma-db-comments-generator
> 
> A big thank you to @Jyrno42

**Comment by thelinuxlich** [comment]
I've also created a option for MySQL, fixing some things on @Jyrno42 code:

https://github.com/thelinuxlich/prisma-mysql-comments-generator

**Comment by yibird** [comment]
I have encountered this problem, do you have any good solutions?

**Comment by charleschan1998** [comment]
support pgSQL as soon as possible. thank uuuuuuuuuuuuuu

**Comment by danhowardms** [comment]
There may be some people reading this who haven't bothered with comments on database columns in the past, and don't really see it as a big deal.  I think once you start using them you'll find they are impossible to live without.

Consider
> price INT
vs
> price INT The base unit price of the product, excluding VAT and discounts, in pence

Without the comment there are so many ways you could make mistakes with this data.

Or

> depot VARCHAR(2)
vs
> depot (VARCHAR(2)) Foreign key to the depot --> depots.code

With the comment, when I'm inspecting the database to troubleshoot some issue, I can write ad-hoc queries and get the JOIN right first without wasting time looking up the relationships.

Or

> desc TEXT
vs
> desc TEXT Long description of the product, in HTML format

With the comment, no-one will accidentally put plain text in this column.

I can't emphasise enough how beneficial database column comments have been for me throughout my career.  A huge +1 from me to add this feature!

**Comment by AnanduKarthi** [comment]
When we pull the exting pg DB, the comments are not in the schema Also, there is a warning in the VS Code, but the CLI validation is successful

**Comment by daedalus28** [comment]
Adding a fresh perspective - it's not for everyone, but if you use an LLM with an MCP server to talk to your database directly, the long tail of work now is building up context in form of documentation about your database (e.g. explaining what fields mean to know when to use them, etc). We can generate context from our prisma files today and just kinda pass it around, but it gets outdated quickly and is more cumbersome. Native DB comments show up when the LLM inspects the database. I'm currently investigating the generator described in this thread, but I'd strongly encourage the prisma team to reconsider this as a priority.

**As LLMs with MCP servers increasingly query databases directly, native DB comments become critical for providing context**. Maybe this can land now that we can tie this to the AI hype train 😄

**Comment by cw1997** [comment]
Excuse, is there any updates on this? This issue has been around for 5 years. I think it is very important now because AI might leverage database schemas for insight information.

**Comment by xlordz** [comment]
> When we pull the exting pg DB, the comments are not in the schema Also, there is a warning in the VS Code, but the CLI validation is successful

🎯 I always used to create my ER Models outside code and don't use migrations in IDE (use external separated migrations) because our systems are always bigger than just front/backend. DBA is DBA and Dev is Dev. And our models are always very documented.

🚀 A must is importing the comments when we prisma:pull because all hard work is already made by this fantastic tool, why not just read the comments and create the triple slash in prisma model ?

**Comment by Ansh714** [comment]
This could be a huge unlock!!

**Comment by zacharyhansen** [comment]
Also bumping this - would be huge for LLM context

**Comment by Vaviloff** [comment]
Fresh Prisma placed this into schema after `db pull`:
```
/// This model or at least one of its fields has comments in the database, and requires an additional setup for migrations: Read more: https://pris.ly/d/database-comments
```
So what's the "setup" currently?

**Comment by melroy89** [comment]
If I want to migrate my existing database to prisma. I would now lose all my comments in the columns and tables.

How could this still not exists in Prisma!??

**Comment by thelinuxlich** [comment]
@melroy89 if you aren't using the ORM, just the schema declaration, you can use another library, like https://github.com/thelinuxlich/mutano

**Comment by melroy89** [comment]
> [@melroy89](https://github.com/melroy89) if you aren't using the ORM, just the schema declaration, you can use another library, like https://github.com/thelinuxlich/mutano

Thanks. Do you also have native mariadb support. It's not explicitly listed.

I don't use the orm part. But I do want to use the migrations part. So it should generate sql files etc. I don't see any destination towards sql. Since I'm talking table and column comments from Prisma schema to sql.

**Comment by thelinuxlich** [comment]
I didn't test with MariaDB but I'm going to, thanks for the suggestion.
